File: gemini/gemini_agent.py
import asyncio
import os
import sys
from typing import Any, List
from dotenv import load_dotenv
from google.genai import Client, types
from google.genai.types import GenerateContentConfig, GenerateContentResponse

load_dotenv()

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../ollama")))
# pyrefly: ignore [missing-import]
from ollama_tools import (
    fetch_website_text,
    list_files,
    read_text_files,
    read_image_file,
    create_file,
    create_text_file,
    get_video_screenshot,
    get_target_info,
    edit_text_files
)

available_functions = {
    'fetch_website_text': fetch_website_text,
    'list_files': list_files,
    'read_text_files': read_text_files,
    'read_image_file': read_image_file,
    'create_file': create_file,
    'create_text_file': create_text_file,
    'get_video_screenshot': get_video_screenshot,
    'get_target_info': get_target_info,
    'edit_text_files': edit_text_files,
}

async def gemini_agent(
    client: Client, messages: List[dict], tools: List[Any], model: str, system_instruction: str = None
) -> GenerateContentResponse:
    config = GenerateContentConfig(
        system_instruction=system_instruction,
        tools=tools,
        temperature=0.5,
    )
    return await client.aio.models.generate_content(
        model=model, contents=messages, config=config
    )

def system_prompt() -> str:
    return (
        "You are a personal AI assistant running locally. "
        "You are given tools to help you with your tasks, use them when necessary. "
        "Your tools calls are listed as you call them, so you don't lose track of them. "
        "It is thus nice to add a 'plan' to your response, so you don't lose track of what's important. "
        "Such 'scratchpad' is added to your chat context as your turn's response."
    )

async def main():
    api_key = os.getenv("GEMINI_API_KEY")
    if not api_key:
        api_key = input("Enter your Gemini API Key: ")
    client = Client(api_key=api_key)

    instruction = system_prompt()
    tools = [
        fetch_website_text,
        list_files,
        read_text_files,
        read_image_file,
        create_file,
        create_text_file,
        get_video_screenshot,
        get_target_info,
        edit_text_files
    ]
    messages = []

    print("\nChatbot initialized. Type your message below, or '/exit' to quit.\n")

    while True:
        user_input = input("\n📝 You: ")
        if user_input.strip().lower() == "/exit":
            break
        if not user_input.strip():
            continue
            
        messages.append({"role": "user", "parts": [{"text": user_input}]})
        
        max_loop_limit = 16
        for loop_counter in range(max_loop_limit):
            try:

                custom_safety_settings = [
                    {
                        "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
                        "threshold": "BLOCK_NONE",
                    },
                ]
                response = await gemini_agent(
                    client,
                    messages, 
                    tools, 
                    "gemini-3.1-flash-lite-preview", 
                    system_instruction=instruction,
                    safety_settings=custom_safety_settings
                )
                
                usage = response.usage_metadata
                if usage:
                    print(f"\n📊 Tokens: {usage.total_token_count} (Input: {usage.prompt_token_count} | Output: {usage.candidates_token_count})")
                    
                if response.text:
                    print(f"\n🤖 Assistant: {response.text}\n")
                    
                messages.append({"role": "model", "parts": response.candidates[0].content.parts})
                
                if response.function_calls:
                    tool_parts = []
                    images_to_attach = []
                    
                    for tool_call in response.function_calls:
                        function_name = tool_call.name
                        arguments = tool_call.args
                        
                        print(f"\n{function_name} is being executed...")
                        
                        if function_name in available_functions:
                            try:
                                result = available_functions[function_name](**arguments)
                            except Exception as e:
                                result = f"Error executing tool: {str(e)}"
                        else:
                            result = f"Error: Tool {function_name} not found."
                            
                        function_response_part = types.Part.from_function_response(
                            name=function_name,
                            response={"result": str(result)}
                        )
                        tool_parts.append(function_response_part)
                        
                        if function_name == 'read_image_file' and "Success" in str(result):
                            image_paths = arguments.get('image_paths') or arguments.get('image_path') or []
                            if isinstance(image_paths, str):
                                image_paths = [image_paths]
                            for path in image_paths:
                                abs_path = os.path.abspath(path)
                                if f"Success: Loaded image from '{abs_path}'." in result:
                                    try:
                                        with open(abs_path, 'rb') as f:
                                            img_bytes = f.read()
                                        _, ext = os.path.splitext(abs_path)
                                        ext = ext.lower()
                                        mime_type = 'image/jpeg'
                                        if ext == '.png':
                                            mime_type = 'image/png'
                                        elif ext == '.bmp':
                                            mime_type = 'image/bmp'
                                        elif ext == '.webp':
                                            mime_type = 'image/webp'
                                            
                                        images_to_attach.append(types.Part.from_bytes(
                                            data=img_bytes,
                                            mime_type=mime_type,
                                        ))
                                    except Exception as e:
                                        print(f"Failed to load image bytes for {abs_path}: {e}")
                                        
                    messages.append({"role": "user", "parts": tool_parts})
                    
                    if images_to_attach:
                        messages.append({
                            "role": "user",
                            "parts": images_to_attach + [{"text": "Here are the images you requested to read."}]
                        })
                else:
                    break
                    
            except Exception as e:
                print(f"\n[Error querying Gemini]: {e}")
                break

if __name__ == "__main__":
    asyncio.run(main())

File: typescript/gemini/gemini_agent.ts
import { GoogleGenAI } from "@google/genai";
import * as dotenv from "dotenv";
import * as readline from "readline";
import * as fs from "fs";
import * as path from "path";
import { available_functions, toolDeclarations } from "../tools";

dotenv.config();

const system_prompt = `You are a personal AI assistant running locally.
You are given tools to help you with your tasks, use them when necessary.
Your tools calls are listed as you call them, so you don't lose track of them.
It is thus nice to add a 'plan' to your response, so you don't lose track of what's important.
Such 'scratchpad' is added to your chat context as your turn's response.`;

async function main() {
  let apiKey = process.env.GEMINI_API_KEY;

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  if (!apiKey) {
    apiKey = await new Promise<string>((resolve) => {
      rl.question("Enter your Gemini API Key: ", (answer) => {
        resolve(answer.trim());
      });
    });
  }

  const ai = new GoogleGenAI({ apiKey });

  const tools = [{ functionDeclarations: toolDeclarations }];
  let contents: any[] = [];

  console.log(
    "\nChatbot initialized. Type your message below, or '/exit' to quit.\n",
  );

  const promptUser = () => {
    rl.question("\n📝 You: ", async (userInput) => {
      if (userInput.trim().toLowerCase() === "/exit") {
        rl.close();
        return;
      }
      if (!userInput.trim()) {
        promptUser();
        return;
      }

      contents.push({ role: "user", parts: [{ text: userInput }] });

      const max_loop_limit = 16;

      for (let loopCounter = 0; loopCounter < max_loop_limit; loopCounter++) {
        try {
          const config: any = {
            systemInstruction: { parts: [{ text: system_prompt }] },
            tools: tools,
            temperature: 0.5,
          };

          const response = await ai.models.generateContent({
            model: "gemini-3.1-flash-lite-preview",
            contents: contents,
            config: config,
          });

          // In Typescript SDK, usageMetadata is typically available
          if (response.usageMetadata) {
            const { totalTokenCount, promptTokenCount, candidatesTokenCount } =
              response.usageMetadata;
            console.log(
              `\n📊 Tokens: ${totalTokenCount} (Input: ${promptTokenCount} | Output: ${candidatesTokenCount})`,
            );
          }

          if (response.text) {
            console.log(`\n🤖 Assistant: ${response.text}\n`);
          }

          // Append the model's response to the history
          const candidateContent = response.candidates?.[0]?.content;
          if (candidateContent && candidateContent.parts) {
            contents.push({
              role: "model",
              parts: candidateContent.parts,
            });
          }

          if (response.functionCalls && response.functionCalls.length > 0) {
            const toolParts: any[] = [];
            const imagesToAttach: any[] = [];

            for (const toolCall of response.functionCalls) {
              const functionName = String(toolCall.name);
              const args: any = toolCall.args || {};

              console.log(`\n${functionName} is being executed...`);

              let result: any;
              if (functionName in available_functions) {
                try {
                  // Map args object to function call
                  // We need to pass args object correctly. In JS, usually it's passed as individual args or just the object
                  // For simplicity and based on our definitions, most take individual arguments.
                  // But since arguments can be in any order in the JSON, let's extract them based on known names or just use Object.values?
                  // Wait, the functions we defined accept positional arguments or specific names.
                  // We should adjust how we call them based on the tool.

                  if (functionName === "fetch_website_text")
                    result = await available_functions[functionName](args.url);
                  else if (functionName === "list_files")
                    result = available_functions[functionName](
                      args.directory,
                      args.grep,
                    );
                  else if (functionName === "read_image_file")
                    result = available_functions[functionName](
                      args.image_paths,
                    );
                  else if (functionName === "create_file")
                    result = available_functions[functionName](
                      args.name,
                      args.extension,
                      args.content,
                    );
                  else if (functionName === "create_text_file")
                    result = available_functions[functionName](
                      args.file_path,
                      args.content,
                    );
                  else if (functionName === "get_video_screenshot")
                    result = available_functions[functionName](
                      args.video_path,
                      args.timestamp,
                    );
                  else if (functionName === "get_target_info")
                    result = available_functions[functionName](
                      args.target_path,
                    );
                  else if (functionName === "read_text_files")
                    result = available_functions[functionName](
                      args.file_paths,
                      args.read_by_chunks_of_40,
                    );
                  else if (functionName === "edit_text_files")
                    result = available_functions[functionName](
                      args.file_path,
                      args.chunks,
                    );
                } catch (e: any) {
                  result = `Error executing tool: ${e.message}`;
                }
              } else {
                result = `Error: Tool ${functionName} not found.`;
              }

              toolParts.push({
                functionResponse: {
                  name: functionName,
                  response: { result: String(result) },
                  id: toolCall.id,
                },
              });

              if (
                functionName === "read_image_file" &&
                String(result).includes("Success")
              ) {
                let imagePaths = args.image_paths || args.image_path || [];
                if (typeof imagePaths === "string") imagePaths = [imagePaths];

                for (const p of imagePaths) {
                  const absPath = path.resolve(p);
                  if (
                    String(result).includes(
                      `Success: Loaded image from '${absPath}'`,
                    )
                  ) {
                    try {
                      const imgBytes = fs.readFileSync(absPath);
                      const ext = path.extname(absPath).toLowerCase();
                      let mimeType = "image/jpeg";
                      if (ext === ".png") mimeType = "image/png";
                      else if (ext === ".bmp") mimeType = "image/bmp";
                      else if (ext === ".webp") mimeType = "image/webp";

                      imagesToAttach.push({
                        inlineData: {
                          data: imgBytes.toString("base64"),
                          mimeType: mimeType,
                        },
                      });
                    } catch (e: any) {
                      console.log(
                        `Failed to load image bytes for ${absPath}: ${e.message}`,
                      );
                    }
                  }
                }
              }
            }

            contents.push({ role: "user", parts: toolParts });

            if (imagesToAttach.length > 0) {
              contents.push({
                role: "user",
                parts: [
                  ...imagesToAttach,
                  { text: "Here are the images you requested to read." },
                ],
              });
            }
          } else {
            break;
          }
        } catch (e: any) {
          console.log(`\n[Error querying Gemini]: ${e.message}`);
          break;
        }
      }

      promptUser();
    });
  };

  promptUser();
}

main().catch(console.error);
